[AIRFLOW-3892] Create Redis pub sub sensor

mans2singh opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-3892/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3892
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:  Create sensor for reading messages from Redis.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
   -  ` tests/contrib/sensors/test_redis_pub_sub_sensor.py`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

mans2singh committed:
[AIRFLOW-3892] Create Redis pub sub sensor

zhongjiajie commented on tests/contrib/sensors/test_redis_pub_sub_sensor.py:
```diff
@@ -0,0 +1,90 @@
+from mock import patch
+from mock import call
+
+from unittest.mock import MagicMock
```
@mans2singh FYI, CI failed due to `from unittest.mock import MagicMock`. Maybe you could do
```suggestion
from mock import MagicMock
```
and than optimize imports

mans2singh replied:
Corrected.

zhongjiajie commented on airflow/contrib/sensors/redis_pub_sub_sensor.py:
```diff
@@ -0,0 +1,72 @@
+        :param context: the context object
+        :type context: dict
+        :return: ``True`` if message (with type 'message') is available or ``False`` if not
+        :rtype: bool
```
Maybe we could remove `rtype`
```suggestion
```

mans2singh replied:
Removed from doc.

zhongjiajie commented on airflow/contrib/sensors/redis_pub_sub_sensor.py:
```diff
@@ -0,0 +1,72 @@
+    ui_color = '#f0eee4'
+
+    @apply_defaults
+    def __init__(self, channels, hook, *args, **kwargs):
```
I think we should use `redis_conn_id` instand of `hook`. we already have `redis_hook` in `airflow.contrib.hook.redis_hook` we could use it.
```suggestion
    def __init__(self, channels, redis_conn_id, *args, **kwargs):
```
than we could create connection using 
```python
self.hook = RedisHook(redis_conn_id=self.redis_conn_id)
```

mans2singh replied:
Hi @zhongjiajie - I was trying the decouple the responsibility of creating the hook from using it by the sensor . This will allow injection of the hook and use mock hooks for unit testing.  
Let me know your thoughts.

zhongjiajie replied:
@mans2singh I think we should use conn str rather than hook as parameter. airflow provide function to create hook by conn str.

mans2singh replied:
@zhongjiajie - 
I have seen that pattern of passing a connection id and the sensor/operator creating the hook and using it.   
I was trying to use dependency injection pattern to make it easy to inject a mock hook and test the sensor.  
Is there any disadvantage of using this pattern ?
Thanks for your advice.

zhongjiajie replied:
@mans2singh disadvantage is we have to write code to create hook in DAG file. I think not high cohesion

If we pass conn_id str, user could just pass conn_id str to active sensors. For unittest I think we could do like other sensors.

mans2singh replied:
Changed to use redis_conn_id as recommended.

zhongjiajie commented on tests/contrib/sensors/test_redis_pub_sub_sensor.py:
```diff
@@ -0,0 +1,90 @@
+        configuration.load_test_config()
+
+        self.log = logging.getLogger()
+        self.log.setLevel(logging.DEBUG)
```
I think we could do unittest without log
```suggestion
```

mans2singh replied:
Removed.

zhongjiajie reviewed (commented):
Left some review here.

mans2singh committed:
[AIRFLOW-3892] - Updated based on review comments (#4712)

Use redis_conn_id, removed logging from test, corrected unittest import, removed rtype from poke doc

zhongjiajie commented on airflow/contrib/sensors/redis_pub_sub_sensor.py:
```diff
@@ -0,0 +1,76 @@
+        super(RedisPubSubSensor, self).__init__(*args, **kwargs)
+        self.channels = channels
+        self.redis_conn_id = redis_conn_id
+        self.pubsub = RedisHook(redis_conn_id=self.redis_conn_id).get_conn().pubsub()
```
Should we move L47 and L48 into function `poke` before `message = self.pubsub.get_message()`? discuss

mans2singh replied:
I placed these lines in the `__init__` function so that they are executed once and `poke`, which can be called multiple times, can just check the message.

zhongjiajie replied:
@mans2singh Ok, I get your point. maybe we should ask advice from maintainers about it. you can ask their reivew now

zhongjiajie commented on airflow/contrib/sensors/redis_pub_sub_sensor.py:
```diff
@@ -0,0 +1,76 @@
+        self.log.info('RedisPubSubSensor checking for message on channels: %s', self.channels)
+
+        message = self.pubsub.get_message()
+        self.log.info('Message %s from channel %s', message, self.channels)
```
In L66 check is `message` is None, but log message here. will this get `Message None from channel %s` or not? And L68 you log message and channel again. Is L63 necessary?

mans2singh replied:
There is some redundancy in the logs but it is easier to review the logs the see that the data/channel when the message is received.  I can remove it if required.

zhongjiajie replied:
@mans2singh Maybe we should discuss with maintainers. You could ask their advice while we finish this review.

mans2singh replied:
@zhongjiajie - I've removed the log as per your recommendation.  Thanks for your review.

zhongjiajie commented on tests/contrib/sensors/test_redis_pub_sub_sensor.py:
```diff
@@ -0,0 +1,80 @@
+from airflow.utils import timezone
+from mock import patch
+from mock import call
+from mock import MagicMock
```
should we use single line
```suggestion
from mock import MagicMock patch call
```

mans2singh replied:
Done.

zhongjiajie reviewed (commented):
Please find my two cents

mans2singh committed:
[AIRFLOW-3892] Combined import based on review comments

mans2singh committed:
[AIRFLOW-3892] Removed extra log based on review comments

Fokko commented on tests/contrib/sensors/test_redis_pub_sub_sensor.py:
```diff
@@ -0,0 +1,77 @@
+
+class TestRedisPubSubSensor(unittest.TestCase):
+
+    @patch('airflow.contrib.hooks.redis_hook.RedisHook.get_conn')
```
In the CI actually runs a Redis using docker-compose. Maybe you can do end to end testing instead of mocking. This makes it easier when you want to test against newer versions of Redis. What do you think?

mans2singh replied:
@Fokko - 

I see that the redis key sensor uses an integration test and will try your recommendation.  

Thanks.

mans2singh committed:
[AIRFLOW-3892] - Added integration tests based on review comments

mans2singh committed:
[AIRFLOW-3892] - Added example dag to kick off the build

mans2singh committed:
[AIRFLOW-3892] - Removed example dag based on review comments
